### PR TITLE
feat(broadcastQueryClient): experimental support for tab/window syncing

### PR DESCRIPTION
**Pull request by tannerlinsley**

### CODE REVIEW

**Comment by tannerlinsley** [comment]
This PR relies on #1792

File: src/broadcastQueryClient-experimental/index.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      return
+    }
+
+    const { type, queryHash, queryKey, state } = action
+
+    if (type === 'queryUpdated') {
+      const query = queryCache.get(queryHash)
+
+      if (query) {
+        query.setState(state, {
```
Comment (boschni) [inline src/broadcastQueryClient-experimental/index.ts:60]: We might also be able to use a flag to indicate a transaction:
```ts
function broadcastQueryClient() {
  let transaction = false
  
  queryClient.getQueryCache().subscribe((query, event) => {
    if (transaction) {
      return
    }
  })
  
  channel.onmessage = action => {
    transaction = true
    query.setState(state)
    transaction = false
  }
}
```

File: docs/src/pages/plugins/broadcastQueryClient.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+broadcastQueryClient({ queryClient, broadcastChannel })
+```
+
+### `Options`
+
+An object of options:
+
+```ts
+interface PersistQueryClientOptions {
+  /** The QueryClient to persist */
```
Comment (TkDodo) [inline docs/src/pages/plugins/broadcastQueryClient.md:45]: I think this is still copy-pated from the persist plugin and should be `BroadcastQueryClientOptions` here :)

**Review by TkDodo — COMMENTED** [review]
General question: It seems that we are not syncing query-clients, but queries inside the query-cache? If that's the case, can we make this a bit more obvious in the docs? Settings like `queryDefaults` on the client would _not_ be synced as far as I can tell.

Also: Does this work if I have two apps that use the same `broadcastChannel`? If so, do both apps need to call `broadcastQueryClient`, or is this meant for the same app opened in multiple browsers only?

File: src/broadcastQueryClient-experimental/index.ts (hunk 1 above)
Comment (tannerlinsley) [reply src/broadcastQueryClient-experimental/index.ts]: Great idea, done.

File: docs/src/pages/plugins/broadcastQueryClient.md (hunk 2 above)
Comment (tannerlinsley) [reply docs/src/pages/plugins/broadcastQueryClient.md]: Fixed 👍

Comment (tannerlinsley) [reply docs/src/pages/plugins/broadcastQueryClient.md]: To answer your questions:
- **For now**, we are only syncing the cache within the clients, but similar to the persistence plugins, we could include other things as well like mutations and settings like you mentioned.
- Broadcast channels (and all of its fallback polyfills) are restricted by CORS, so this plugin is meant to sync between similar instance of the same app on the same subdomain/CORS. Anything more than that would require websockets and a common server connection, but that's a whole other topic :)

File: docs/src/pages/plugins/broadcastQueryClient.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+```ts
+broadcastQueryClient({ queryClient, broadcastChannel })
+```
+
+### `Options`
+
+An object of options:
+
+```ts
+interface PersistQueryClientOptions {
```
Comment (TkDodo) [inline docs/src/pages/plugins/broadcastQueryClient.md:44]: it still says `PersistQueryClientOptions` here 😅 

```suggestion
interface BroadcastQueryClientOptions {
```

**Review by TkDodo — APPROVED** [review]

**Pull request merged**